pkg/contractreader/extended.go

```diff
@@ -261,10 +268,25 @@ func (e *extendedContractReader) Bind(ctx context.Context, allBindings []types.B
 	e.mu.Lock()
 	defer e.mu.Unlock()
 	for _, binding := range validBindings {
-		e.contractBindingsByName[binding.Name] = append(e.contractBindingsByName[binding.Name], ExtendedBoundContract{
-			BoundAt: time.Now(),
-			Binding: binding,
-		})
+		if e.multiBindAllowed[binding.Name] {
+			e.contractBindingsByName[binding.Name] = append(e.contractBindingsByName[binding.Name], ExtendedBoundContract{
+				BoundAt: time.Now(),
+				Binding: binding,
+			})
```

Comment:
style nit: I would be tempted to rewrite this to avoid the `else`. Something like
```
    // unbind previous contract if needed
    if len(e.contractBindingsByName[binding.Name]) > 0 && e.multiBindAllowed[binding.Name] {
        e.reader.Unbind(...)
        e.contractBindingsByName[binding.Name] = nil
    }

    // existing logic here
}
```